### PR TITLE
ROX-17682, ROX-17684: Download telemetry key

### PR DESCRIPTION
**Pull request by 0x656b694d**
## Description

Release version central binary will download a json document from a URL, defined via `ROX_TELEMETRY_CONFIG_URL` (with some hardcoded default value) on startup if there's no telemetry key defined. Development version will download the config if only a key is also defined via environment variable and if the key is equal to the one found in the config.

Example document:
```json
{ "storage_key_v1": "asdfm52347fhvsdkj" }
```

*Update*: until the default endpoint is implemented, we'll use the hardcoded key.

## Checklist
- [x] Investigated and inspected CI test results
- [ ] Unit test and regression tests added
- [ ] ~Evaluated and added CHANGELOG entry if required~
- [ ] ~Determined and documented upgrade steps~
- [ ] ~Documented user facing changes (create PR based on [openshift/openshift-docs](https://github.com/openshift/openshift-docs) and merge into [rhacs-docs](https://github.com/openshift/openshift-docs/tree/rhacs-docs))~

If any of these don't apply, please comment below.

## Testing Performed

With an HTTP server pod, serving `config.json`, central logs:
```
...
telemetry/centralclient: 2023/06/13 15:08:16.374527 instance_config.go:86: Info: Telemetry configuration has been downloaded from http://10.244.0.18:8000/config.json
...
```

### CODE REVIEW

**Comment by 0x656b694d** [comment]
Current dependencies on/for this PR:
* master
  * **PR #6420** <a href="https://app.graphite.dev/github/pr/stackrox/stackrox/6420" target="_blank">Graphite</a>  👈

This comment was auto-generated by [Graphite](https://app.graphite.dev/github/pr/stackrox/stackrox/6420?utm_source=stack-comment).

File: central/telemetry/centralclient/instance_config.go [hunk 1]
```diff
@@ -33,9 +38,56 @@ var (
 	enabled  bool
 )
 
+const (
+	telemetryConfigURL = "https://telemetry.stackrox.io/config.json"
```
Comment (janisz) [inline central/telemetry/centralclient/instance_config.go:42]: > anonymous access is forbidden

File: central/telemetry/centralclient/instance_config.go [hunk 2]
```diff
@@ -33,9 +38,60 @@ var (
...
+}
+
 func getInstanceConfig() (*phonehome.Config, map[string]any, error) {
 	key := env.TelemetryStorageKey.Setting()
-	if key == "" || env.OfflineModeEnv.BooleanSetting() {
+	if key == disabledKey || env.OfflineModeEnv.BooleanSetting() {
+		return nil, nil, nil
+	}
+
+	if env.TelemetryConfigURL.Setting() != "" &&
```
Comment (janisz) [inline central/telemetry/centralclient/instance_config.go:79]: how about splitting condition. I understand we can return if `env.TelemetryConfigURL.Setting() == ""`

File: central/telemetry/centralclient/instance_config.go [hunk 3]
```diff
@@ -33,9 +38,60 @@ var (
...
+		if err != nil {
+			return nil, nil, err
+		}
+		// The key has to match for development versions.
+		if remoteCfg != nil && (isReleaseBuild() || remoteCfg.Key == key) {
+			key = remoteCfg.Key
+			log.Info("Telemetry configuration has been downloaded from ",
+				env.TelemetryConfigURL.Setting())
+		}
+	}
```
Comment (janisz) [inline central/telemetry/centralclient/instance_config.go:92]:
```suggestion
	key := env.TelemetryStorageKey.Setting()
	configUrl := env.TelemetryConfigURL.Setting()
	if key == disabledKey || env.OfflineModeEnv.BooleanSetting() {
		return nil, nil, nil
	}
	if key == "" && !isReleaseBuild() {
		return nil, nil, nil
	}
	if configUrl == "" {
		return nil, nil, nil
	}
	remoteCfg, err := downloadConfig(configUrl)
	if err != nil {
		return nil, nil, err
	}
	// The key has to match for development versions.
	if remoteCfg != nil && (isReleaseBuild() || remoteCfg.Key == key) {
		key = remoteCfg.Key
		log.Info("Telemetry configuration has been downloaded from ", configUrl)
	}
```

File: central/telemetry/centralclient/instance_config.go [hunk 4]
```diff
@@ -33,9 +38,60 @@ var (
...
+type remoteConfig struct {
+	Key string `json:"storage_key_v1,omitempty"`
+}
+
+func downloadConfig(u string) (*remoteConfig, error) {
+	req, err := http.NewRequest(http.MethodGet, u, nil)
+	if err != nil {
+		return nil, errors.Wrap(err, "cannot request telemetry configuration")
+	}
+	client := http.Client{
```
[the remark this answers is not in the document]
Comment (0x656b694d) [reply central/telemetry/centralclient/instance_config.go:54]: This is to be called once on startup.

File: central/telemetry/centralclient/instance_config.go [hunk 5]
```diff
@@ -33,9 +38,60 @@ var (
...
+	resp, err := client.Do(req)
+	if err != nil {
+		return nil, errors.Wrap(err, "cannot download telemetry configuration")
+	}
+	var cfg *remoteConfig
+	err = json.NewDecoder(resp.Body).Decode(&cfg)
+	return cfg, errors.Wrap(err, "cannot decode telemetry configuration")
+}
+
+func isReleaseBuild() bool {
```
[the remark this answers is not in the document]
Comment (0x656b694d) [reply central/telemetry/centralclient/instance_config.go:67]: It is replaced in other PR, where `version.IsReleaseBuild()` is introduced and used in other places.

File: central/telemetry/centralclient/instance_config.go [hunk 6]
```diff
@@ -33,9 +38,93 @@ var (
...
+		!strings.Contains(version.GetMainVersion(), "-")
+}
+
+// toDownload decides if a configuration with the key need to be downloaded.
+// We want to prevent accidental use of the production key, but still allow
+// developers to test the functionality. So download will only happen for
+// development installations if both a key and an URL are provided. For release
+// versions the key may be empty.
+// See unit tests for the examples.
+func toDownload(isRelease bool, key, cfgURL string) bool {
```
Comment (janisz) [inline central/telemetry/centralclient/instance_config.go:80]: ❤️

**Review by janisz — APPROVED** [review]

**Pull request merged**